Vertical Navigation

mklanjsek opened the pull request:
After ironing a few layout issues with @rossfenrick, I think this is a got point to be reviewed and merged to master. Vertical navigation now works for both expanded and collapsed state with the hand-rolled flyout menu. Unfortunately, there was some history lost during merge from the old branch due to so many changes since that work started.  

@vmware-tanzu/octant-reviewers feedback please!

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/container/container.component.scss:
```diff
@@ -2,50 +2,54 @@
   }
 
   .content-container {
+    margin-bottom: 1.5rem;
```
Why the the content-container need a margin here?

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/content/content.component.html:
```diff
@@ -1,4 +1,4 @@
-<div class="content-component">
+<div clrFocusOnViewInit #scrollTarget class="content-component">
```
`scrollTarget` was removed as part of slider cleanup in https://github.com/vmware-tanzu/octant/commit/bcf1148c0f23811866969e124464a0ba3b093fd7#diff-4399b3274650f85b70c641d7e49db84d

mklanjsek replied:
Removed

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/navigation/navigation.component.html:
```diff
@@ -1,58 +1,106 @@
+                   [routerLinkActiveOptions]="{exact: true}"
+                   [routerLink]="formatPath(section.path)"
+                   routerLinkActive="active"
+                   (mouseover)="flyoutIndex= -1; lastSelection= i"
```
Upon moving the mouse away from the collapsed nav, the item should not remain highlighted

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/navigation/navigation.component.html:
```diff
@@ -1,58 +1,106 @@
+
+  <!-- Theme switcher -->
+  <div class="theme-switcher">
+    <app-theme-switch-button [collapsed]="collapsed"></app-theme-switch-button>
```
nit: The theme switch button in a collapsed navigation should remain on the bottom rather than shifting upwards as a result of collapsed content.

mklanjsek replied:
This was bugging me too, finally figured out how to do it

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/navigation/navigation.component.scss:
```diff
@@ -2,18 +2,147 @@
+  .flyout {
+    display: flex;
+    border: 1px solid var(--navigation-border-color);
+    border-left-width: 0px;
```
It looks odd without a left border

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/namespace/namespace.component.html:
```diff
@@ -1,10 +1,27 @@
+  </ng-template>
+  <ng-template #hideDropdown>
+    <clr-icon shape="namespace"></clr-icon>
+    [All Namespaces]
```
Hit a case where navigation from `/cluster-overview` to `/` via top left octant logo results in stuck on `All Namespaces`

mklanjsek replied:
Done

GuessWhoSamFoo commented on web/src/app/modules/sugarloaf/components/smart/namespace/namespace.component.html:
```diff
@@ -1,10 +1,27 @@
+  <ng-template #hideDropdown>
+    <clr-icon shape="namespace"></clr-icon>
+    [All Namespaces]
+  </ng-template>
```
Another case where being in `/cluster-overview` and expanding a namespaced navigation section changes the namespace.

mklanjsek replied:
I think this is OK - showing the namespace properties is not equivalent to switching to that namespace.

mklanjsek replied:
At the end I decided to follow your recommendation and switch the namespace in this case.

GuessWhoSamFoo commented:
As discussed in slack, we should merge this to master and break apart pending issues into an epic with the main one being losing a level of hierarchy for plugins.

mklanjsek commented:
All good points, thanks @GuessWhoSamFoo!

rossfenrick commented:
>  The sidebar should be collapsed by default. (we may even consider not allowing expansion at all later)

I'd highly recommend against this. A widely accepted UX best practice is that primary navigation should only be hidden if it the viewport is constrained by size (mobile).

bryanl commented:
@rossfenrick The idea is to make it act more like vscode's sidebar. It would not be hidden, but would instead take up less room.

rossfenrick commented:
@bryanl Sorry, when I say "hidden" I mean collapsed/only icons. Something we could change is if the user chooses to collapse the navigation, it stays like that on refresh and the next time they start the app. 

What's the benefit to removing the ability to expand & collapse?

mklanjsek commented:
Expanded/collapsed state should certainly be a user level setting stored in browser local storage, similar to theme and other preferences.

As for default state, expanded state is certainly geared towards new and novice users, so from that point of view it's probably a good choice to be a default state. If we can remember that state once a user flips the switch, we should be able to handle wide spectrum of user affinities.

wwitzel3 commented:
I see lots of tools and other products that don't even offer an expanded state. They use tooltips / etc.. and our flyout menu + tooltip provide all of the information a person needs to navigate and they have a set of medium/large icons for the navigation.

I really like the idea of moving away from the expanded navigation all together, similar to tools like VS code. Making the collapsed view the only view and making the icons bigger.

mklanjsek commented:
I suggest we switch to collapsed, test that internally and make final decision before the next release.  Anybody against that?

mklanjsek commented:
I addressed all above issues, except this one:

> Namespace Overview and Cluster Overview both just read Overview and they are ambiguous when the menu is collapsed.

Unfortunately client does not have enough info to show that in tooltip. Change will be required on backend to add tooltip text, so I'll open a separate issue to handle that..
